code/datums/shuttle.dm

```diff
@@ -0,0 +1,510 @@
+#define NO_TRANSIT 0 //Don't use transit areas
+#define TRANSIT_ACROSS_Z_LEVELS 1 //Only use transit areas if moving to another z-level
+#define TRANSIT_ALWAYS 2 //Always use transit areas
+
+#define LINK_FREE 0
+#define LINK_PASSWORD_ONLY 1
+#define LINK_FORBIDDEN 2
+
+#define COLLISION_DESTROY 0
+#define COLLISION_DISPLACE 1 //this is same as COLLISION_DISPLACE as for now
+
+/datum/shuttle
+	var/name = "shuttle"
+
+	var/list/docking_ports = list() //List of ALL docking ports the shuttle can move to
+
+	var/area/linked_area //the area which contains the docking port and which is moved
+
+	var/obj/structure/docking_port/shuttle/linked_port
+
+	var/obj/structure/docking_port/destination/current_port //where we are at now
+
+	var/obj/structure/docking_port/destination/transit_port //transit port
+
+	var/obj/structure/docking_port/destination/destination_port //where we are moving
+
+	var/list/docking_ports_aboard = list()	//List of ALL docking ports on the shuttle. The shuttle only moves docking ports that are in this list. Setup on initialize()
+											//This exists to stop a shuttle from moving on top of a docking port and then moving it away
+
+	var/use_transit = TRANSIT_ACROSS_Z_LEVELS
+
+	var/dir = NORTH //Direction of the shuttle
+	var/can_rotate = 0 //Whether the shuttle can rotate. This feature is B R O K E N
+
+	var/pre_flight_delay = 100 //This is the time it takes for the shuttle to depart (if there's a transit area) or to travel (if there are no transit areas)
+
+	var/transit_delay = 100	//If there is a transit port, this is the time the shuttle spends in it
+							//If there isn't a transit port, this has no effect. Use the pre_flight_delay var instead
+
+	var/moving = 0 //If the shuttle is currently moving
+
+	var/list/cant_leave_zlevel = list(
+		/obj/item/weapon/disk/nuclear = "The nuclear authentication disk can't be transported on a shuttle.",
+		)
+
+	var/last_moved = 0
+	var/cooldown = 100
+
+	var/innacuracy = 0	//When the shuttle moves, coordinates of its final location will be offset by rand(-innacuracy, innacuracy)
+						//Oh god why
+
+	var/password = 28011
+	var/can_link_to_computer = LINK_FORBIDDEN
+
+	var/collision_type = COLLISION_DESTROY //Whether the shuttle gibs or displaces stuff
+
+	var/obj/machinery/computer/shuttle_core/core_computer
+	var/list/control_consoles = list()
+
+	var/lockdown = 0
+
+/datum/shuttle/New(var/area/starting_area)
+	.=..()
+	shuttles |= src
+
+	if(starting_area)
+		if(ispath(starting_area))
+			linked_area = locate(starting_area)
+		else if(isarea(starting_area))
+			linked_area = starting_area
+		else
+			world.log << "WARNING: Unable to find area [starting_area] in world - [src.type] ([src.name]) will be broken."
```

Comment:
You can use the proc `warning()` here to automatically put the `WARNING: ` and output it to `world.log`